messages/protobuf/impl.go

```diff
@@ -34,14 +34,20 @@ func (*impl) NewFromBinary(data []byte) (messages.Message, error) {
 		return nil, fmt.Errorf("failed to unmarshal message wrapper: %s", err)
 	}
 
-	switch msg.Type.(type) {
+	switch t := msg.Type.(type) {
+	case *Message_Request:
```

Comment:
This function switches based on message type extracted from raw message data, but it's seems that the sealed interface `message.IsRequest` is not used here.

https://github.com/sergefdrv/minbft/blob/8245dae27eba1a6bdb4592aac367daec48ba92b8/messages/protobuf/request.go#L22-L25

`protobuf.request` embeds `messages.IsRequest`, so I guess that you realize the new thing on this type. So if you explain how it's used, the concept will get clearer.

Reply:
That's what I try to convey in [the "explicit interfaces" proposal](https://github.com/golang/go/issues/34996): The sealed interfaces is a tricky solution to the problem.

Here, `msg` is a `Message` structure which is defined by auto-generated code. Protobuf's code generator `protoc` defines a `oneof` field `Message.Type` as an unexported interface `isMessage_Type` with a single unexported method `isMessage_Type()`. The whole purpose of `isMessage_Type` interface is to choose which concrete types can be assigned to the field. This is close to the sealed interfaces construction, but within the same Go package.

Thus, this type switch determines the unmarshaled message type relying on protobuf's definitions. This should work as expected, because the switch cases match concrete types (structures, not interfaces) and `protoc` guarantees the expected semantic of such type switches/assertions. 

However, the returned value of this function is an abstract interface `messages.Message`, defined in `messages/api.go`. That value's type should embed one of `IsXxx` structures defined in `messages/api.go`. The value returned by `newRequest()` in the following line is of `request` type, which embeds `messages.IsRequest`. This embedding ensures that `request` implements the unexported methods of `messages.Request` (the only purpose of `messages.IsRequest`). The remaining methods of `messages.Request` are implemented by `request` itself so that it fulfils `messages.Request` interface.